File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/tools/speedtest/LANSpeedTestViewModel.kt
package com.openu60.feature.tools.speedtest

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okio.Buffer
import java.util.concurrent.TimeUnit
import javax.inject.Inject

data class LANSpeedTestState(
    val phase: String = "idle",
    val progress: Float = 0f,
    val liveSpeedMbps: Double = 0.0,
    val pingMs: Double? = null,
    val downloadMbps: Double? = null,
    val uploadMbps: Double? = null,
    val error: String? = null,
    val isRunning: Boolean = false,
)

@HiltViewModel
class LANSpeedTestViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(LANSpeedTestState())
    val state: StateFlow<LANSpeedTestState> = _state.asStateFlow()

    private var testJob: Job? = null

    private val testSize = 20_000_000

    private val lanClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .build()

    fun startTest() {
        if (_state.value.isRunning) return
        _state.value = LANSpeedTestState(isRunning = true)

        testJob = viewModelScope.launch {
            try {
                // Ping
                _state.value = _state.value.copy(phase = "ping")
                val ping = measurePing()
                _state.value = _state.value.copy(pingMs = ping, progress = 0.2f)

                // Download
                _state.value = _state.value.copy(phase = "download", liveSpeedMbps = 0.0)
                val download = measureDownload()
                _state.value = _state.value.copy(downloadMbps = download, progress = 0.6f)

                // Upload
                _state.value = _state.value.copy(phase = "upload", liveSpeedMbps = 0.0)
                val upload = measureUpload()
                _state.value = _state.value.copy(
                    uploadMbps = upload,
                    progress = 1f,
                    phase = "complete",
                    isRunning = false,
                    liveSpeedMbps = 0.0,
                )
            } catch (e: kotlinx.coroutines.CancellationException) {
                _state.value = _state.value.copy(phase = "cancelled", isRunning = false, liveSpeedMbps = 0.0)
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    phase = "error",
                    error = e.message ?: "Unknown error",
                    isRunning = false,
                    liveSpeedMbps = 0.0,
                )
            }
        }
    }

    fun stopTest() {
        testJob?.cancel()
        testJob = null
    }

    private suspend fun measurePing(): Double = withContext(Dispatchers.IO) {
        val url = "${agentClient.baseURL}/api/lan/ping"
        val rtts = mutableListOf<Double>()

        repeat(10) {
            if (!isActive) throw kotlinx.coroutines.CancellationException()
            val request = Request.Builder()
                .url(url)
                .get()
                .apply { agentClient.token?.let { header("Authorization", "Bearer $it") } }
                .build()

            val start = System.nanoTime()
            lanClient.newCall(request).execute().use { response ->
                response.body?.string()
            }
            val rtt = (System.nanoTime() - start) / 1_000_000.0
            rtts.add(rtt)
        }

        rtts.sort()
        rtts[rtts.size / 2]
    }

    private suspend fun measureDownload(): Double = withContext(Dispatchers.IO) {
        val url = "${agentClient.baseURL}/api/lan/download?size=$testSize"
        val request = Request.Builder()
            .url(url)
            .get()
            .apply { agentClient.token?.let { header("Authorization", "Bearer $it") } }
            .build()

        val response = lanClient.newCall(request).execute()
        val source = response.body!!.source()
        val buf = Buffer()
        var received = 0L
        val start = System.nanoTime()

        while (isActive) {
            val read = source.read(buf, 16384)
            if (read == -1L) break
            received += read
            buf.clear()

            val elapsed = (System.nanoTime() - start) / 1_000_000_000.0
            if (elapsed > 0.1) {
                val mbps = received * 8.0 / (elapsed * 1_000_000)
                val fraction = received.toDouble() / testSize
                _state.value = _state.value.copy(
                    liveSpeedMbps = mbps,
                    progress = (0.2f + fraction.toFloat() * 0.4f).coerceAtMost(0.6f),
                )
            }
        }
        response.close()

        val elapsed = (System.nanoTime() - start) / 1_000_000_000.0
        if (elapsed > 0) received * 8.0 / (elapsed * 1_000_000) else 0.0
    }

    private suspend fun measureUpload(): Double = withContext(Dispatchers.IO) {
        val url = "${agentClient.baseURL}/api/lan/upload"
        val uploadData = ByteArray(testSize)
        val body = uploadData.toRequestBody("application/octet-stream".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(body)
            .apply { agentClient.token?.let { header("Authorization", "Bearer $it") } }
            .build()

        val start = System.nanoTime()
        val response = lanClient.newCall(request).execute()
        val elapsed = (System.nanoTime() - start) / 1_000_000_000.0
        val responseBody = response.body?.string() ?: ""
        response.close()

        // Parse server-measured result
        try {
            val element = Json.parseToJsonElement(responseBody)
            val data = element.jsonObject["data"]?.jsonObject
            val serverMbps = data?.get("mbps")?.jsonPrimitive?.doubleOrNull
            if (serverMbps != null) return@withContext serverMbps
        } catch (_: Exception) {}

        // Fallback to client-side
        if (elapsed > 0) testSize * 8.0 / (elapsed * 1_000_000) else 0.0
    }
}

File: mobile/ios/OpenU60/Features/Tools/SpeedTest/LANSpeedTestViewModel.swift
import Foundation

@Observable
@MainActor
final class LANSpeedTestViewModel {
    var phase: String = "idle"
    var pingMs: Double?
    var downloadMbps: Double?
    var uploadMbps: Double?
    var liveSpeedMbps: Double = 0
    var progress: Double = 0
    var isRunning: Bool = false
    var error: String?

    private let client: AgentClient
    private nonisolated(unsafe) var testTask: Task<Void, Never>?

    private let testSize = 20_000_000

    init(client: AgentClient) {
        self.client = client
    }

    deinit {
        testTask?.cancel()
    }

    func startTest() {
        guard !isRunning else { return }
        isRunning = true
        phase = "idle"
        pingMs = nil
        downloadMbps = nil
        uploadMbps = nil
        liveSpeedMbps = 0
        progress = 0
        error = nil

        testTask = Task {
            do {
                try Task.checkCancellation()
                phase = "ping"
                pingMs = try await measurePing()
                progress = 0.2

                try Task.checkCancellation()
                phase = "download"
                liveSpeedMbps = 0
                downloadMbps = try await measureDownload()
                progress = 0.6

                try Task.checkCancellation()
                phase = "upload"
                liveSpeedMbps = 0
                uploadMbps = try await measureUpload()
                progress = 1.0

                phase = "complete"
            } catch is CancellationError {
                phase = "cancelled"
            } catch {
                self.error = error.localizedDescription
                phase = "error"
            }
            liveSpeedMbps = 0
            isRunning = false
        }
    }

    func stopTest() {
        testTask?.cancel()
        testTask = nil
    }

    // MARK: - Ping

    private func measurePing() async throws -> Double {
        let baseURL = client.baseURL
        let token = client.token
        guard let url = URL(string: "\(baseURL)/api/lan/ping") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 5
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }

        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 5
        let session = URLSession(configuration: config)
        defer { session.invalidateAndCancel() }

        var rtts: [Double] = []
        for _ in 0..<10 {
            try Task.checkCancellation()
            let start = CFAbsoluteTimeGetCurrent()
            let (_, _) = try await session.data(for: request)
            let rtt = (CFAbsoluteTimeGetCurrent() - start) * 1000
            rtts.append(rtt)
        }

        guard !rtts.isEmpty else { throw URLError(.cannotConnectToHost) }
        rtts.sort()
        return rtts[rtts.count / 2]
    }

    // MARK: - Download

    private func measureDownload() async throws -> Double {
        let baseURL = client.baseURL
        let token = client.token
        let size = testSize
        guard let url = URL(string: "\(baseURL)/api/lan/download?size=\(size)") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 120
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }

        let measurer = DownloadMeasurer(expectedSize: Int64(size)) { [weak self] mbps, fraction in
            Task { @MainActor [weak self] in
                self?.liveSpeedMbps = mbps
                self?.progress = 0.2 + fraction * 0.4
            }
        }

        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 120
        config.timeoutIntervalForResource = 120
        let session = URLSession(configuration: config, delegate: measurer, delegateQueue: nil)
        defer { session.invalidateAndCancel() }

        let (_, _) = try await session.data(for: request, delegate: measurer)
        return measurer.finalMbps
    }

    // MARK: - Upload

    private func measureUpload() async throws -> Double {
        let baseURL = client.baseURL
        let token = client.token
        let size = testSize
        guard let url = URL(string: "\(baseURL)/api/lan/upload") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 120
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        request.httpBody = Data(count: size)

        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 120
        config.timeoutIntervalForResource = 120
        let session = URLSession(configuration: config)
        defer { session.invalidateAndCancel() }

        let start = CFAbsoluteTimeGetCurrent()
        let (data, _) = try await session.data(for: request)
        let clientElapsed = CFAbsoluteTimeGetCurrent() - start

        // Parse server-measured result (primary)
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let inner = json["data"] as? [String: Any],
           let serverMbps = inner["mbps"] as? Double {
            return serverMbps
        }

        // Fallback to client-side calculation
        return clientElapsed > 0 ? Double(size) * 8.0 / (clientElapsed * 1_000_000) : 0
    }
}

// MARK: - Download delegate

private final class DownloadMeasurer: NSObject, URLSessionDataDelegate, @unchecked Sendable {
    private let startTime = CFAbsoluteTimeGetCurrent()
    private var totalReceived: Int64 = 0
    private let expectedSize: Int64
    private let onProgress: @Sendable (Double, Double) -> Void

    var finalMbps: Double {
        let elapsed = CFAbsoluteTimeGetCurrent() - startTime
        return elapsed > 0 ? Double(totalReceived) * 8.0 / (elapsed * 1_000_000) : 0
    }

    init(expectedSize: Int64, onProgress: @escaping @Sendable (Double, Double) -> Void) {
        self.expectedSize = expectedSize
        self.onProgress = onProgress
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        totalReceived += Int64(data.count)
        let elapsed = CFAbsoluteTimeGetCurrent() - startTime
        guard elapsed > 0.1 else { return }
        let mbps = Double(totalReceived) * 8.0 / (elapsed * 1_000_000)
        let fraction = min(Double(totalReceived) / Double(expectedSize), 1.0)
        onProgress(mbps, fraction)
    }
}

